plugin/src/main/java/com/denizenscript/denizen/scripts/commands/player/ResourcePackCommand.java

```diff
@@ -77,12 +90,44 @@ public static void autoExecute(ScriptEntry scriptEntry,
             Debug.echoError("Invalid resource_pack hash. Should be 40 characters of hexadecimal data.");
             return;
         }
-        for (PlayerTag player : targets) {
-            if (!player.isOnline()) {
-                Debug.echoDebug(scriptEntry, "Player is offline, can't send resource pack to them. Skipping.");
-                continue;
+        if (!NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20) && (action == Actions.ADD || id != null)) {
+            throw new UnsupportedOperationException();
+        }
+        switch (action) {
+            case SET -> {
+                for (PlayerTag player : targets) {
+                    if (!player.isOnline()) {
+                        Debug.echoDebug(scriptEntry, "Player is offline, can't send resource pack to them. Skipping.");
+                        continue;
+                    }
+                    PaperAPITools.instance.setResourcePack(player.getPlayerEntity(), url, hash, forced, prompt, id);
+                }
+            }
+            case ADD -> {
+                byte[] hashData = new byte[20];
+                for (int i = 0; i < 20; i++) {
+                    hashData[i] = (byte) Integer.parseInt(hash.substring(i * 2, i * 2 + 2), 16);
+                }
+                UUID packUUID;
+                if (id == null) {
+                    packUUID = UUID.nameUUIDFromBytes(url.getBytes(StandardCharsets.UTF_8));
+                }
+                else {
+                    try {
+                        packUUID = UUID.fromString(id);
+                    }
+                    catch (IllegalArgumentException e) {
+                        packUUID = UUID.nameUUIDFromBytes(id.getBytes(StandardCharsets.UTF_8));
+                    }
+                }
+                for (PlayerTag player : targets) {
+                    if (!player.isOnline()) {
+                        Debug.echoDebug(scriptEntry, "Player is offline, can't send resource pack to them. Skipping.");
+                        continue;
+                    }
+                    player.getPlayerEntity().addResourcePack(packUUID, url, hashData, prompt, forced);

```

Comment:
Using Spigot API here vs Paper API for setting means that `set` allows advanced text stuff in the prompt and `add` doesn't - might be better to use Paper API here as well for consistency?

Reply:
There isn't a paper api version of this that I saw, other than the Audience stuff. Can you link me what you're referencing?

Reply:
Don't think Paper has a direct alternative to the Spigot method? but you can just use adventure, untested but should be something like this:
```java
ResourcePackRequest.Builder builder = ResourcePackRequest.resourcePackRequest();
builder.packs(ResourcePackInfo.resourcePackInfo(UUID.fromString("123"), URI.create("https://www.pack.com/pack.zip"), "hash-13th0aing"));
builder.prompt(Component.text("Use pack pls").color(NamedTextColor.AQUA));
builder.replace(false);
builder.required(true);
player.sendResourcePacks(builder);
```